File: simple_deflate.rb
class SimpleDeflate
  class << self
    def encode(message)
      encoded = encode_impl(message)
      [encoded[:length]].pack('N') + [ :header1, :header2, :body ].map { |k| encoded[k] }.join
    end

    def decode(message)
      decode_impl(message)
    end

    private

    def build_tree(chars)
        a = chars.uniq.map {|c| {char: c, count: chars.count(c)}}.sort {|a, b| [b[:count], a[:char]] <=> [a[:count], b[:char]]}
        pq = a.map { |c| {char: c[:char], count: c[:count], left: nil, right: nil} }

        while pq.size > 1
            left, right = pq.pop(2)

            if left[:count] < right[:count]
                left, right = right, left
            end

            pq << {char: nil, count: left[:count] + right[:count], left: left, right: right}

            pq.sort! {|a,b| [b[:count], a[:char] || 0] <=> [a[:count], b[:char] || 0]}
        end

        pq[0]
    end

    def build_table(tree)
        q = [{node: tree, code: ''}]

        codes = {}

        while q.size > 0
            e = q.shift

            if e[:node][:left].nil? && e[:node][:right].nil?
                codes[e[:node][:char]] = e[:code]
                next
            end

            q.push({node: e[:node][:left], code: e[:code] + '0'}) if !e[:node][:left].nil?

            q.push({node: e[:node][:right], code: e[:code] + '1'}) if !e[:node][:right].nil?
        end

        codes
    end

    def build_canonical_table(codes)
        codes = codes.sort_by {|k, v| [v.length, k]}
        canonical_codes = {}
        curr_code = 0
        prev_length = 0

        codes.each do |char, old_code|
            curr_code <<= (old_code.length - prev_length)
            canonical_codes[char] = curr_code.to_s(2).rjust(old_code.length, '0')
            curr_code += 1
            prev_length = old_code.length
        end

        canonical_codes
    end

    def running_codes_lengths(codes)
        codes_lengths = (0..255).map {|i| (codes[i] || '').length}

        result = []
        i = 0

        while i < codes_lengths.size
            len = codes_lengths[i]
            run_length = 1

            while i + run_length < codes_lengths.size && codes_lengths[i + run_length] == len
                run_length += 1
            end

            i += run_length

            if len == 0
                while run_length > 0
                    if run_length >= 11
                        # code 18, repeat '0' 11..138 times
                        diff = [run_length, 138].min
                        result << [18, diff - 11]
                        run_length -= diff
                    elsif run_length >= 3
                        # code 17, repeat '0' 3..10 times
                        diff = [run_length, 10].min
                        result << [17, diff - 3]
                        run_length -= diff
                    else
                        result += [0] * run_length
                        run_length = 0
                    end
                end
            elsif len != 0 && run_length >= 3
                result << len
                run_length -= 1

                while run_length >= 3
                    # code 16, repeat previous value 3..6 times
                    diff = [run_length, 6].min
                    result << [16, diff - 3]
                    run_length -= diff
                end
            end

            run_length.times { result << len }
        end

        result
    end

    def encode_impl(s)
        bytes = s.bytes
        tree = build_tree(bytes)
        codes = build_table(tree)
        canonical_codes = build_canonical_table(codes)

        codes_lengths = running_codes_lengths(canonical_codes)
        raw_codes_lengths = codes_lengths.map {|i| if i.is_a?(Array) then i[0] else i end}

        codes_lengths_tree = build_tree(raw_codes_lengths)
        codes_lengths_codes = build_table(codes_lengths_tree)
        canonical_codes_lengths = build_canonical_table(codes_lengths_codes)

        body_bits = bytes.map {|c| canonical_codes[c]}.join
        padded_body_bits = body_bits.ljust((body_bits.length + 7) / 8 * 8, '0')
        encoded_body = padded_body_bits.chars.each_slice(8).map {|slice| slice.join.to_i(2).chr}.join
        encoded_codes_lengths = codes_lengths.map do |c|
            if c.is_a?(Array)
                code, extra = c
                extra_bits = case code
                    when 16 then 2
                    when 17 then 3
                    when 18 then 7
                end
                canonical_codes_lengths[code] + extra.to_s(2).rjust(extra_bits, '0')
            else
                canonical_codes_lengths[c]
            end
        end.join

        bit_string = (0..18).map {|c| (canonical_codes_lengths[c] || '').length.to_s(2).rjust(3, '0') }.join
        padded_bit_string = bit_string.ljust((bit_string.length + 7) / 8 * 8, '0')
        encoded_codes_lengths_tree = padded_bit_string.chars.each_slice(8).map {|slice| slice.join.to_i(2).chr}.join

        padded_encoded_codes_lengths = encoded_codes_lengths.ljust((encoded_codes_lengths.length + 7) / 8 * 8, '0')
        encoded_codes_lengths_body = padded_encoded_codes_lengths.chars.each_slice(8).map {|slice| slice.join.to_i(2).chr}.join

        {
            header1: encoded_codes_lengths_tree,
            header2: encoded_codes_lengths_body,
            body: encoded_body,
            length: bytes.length
        }
    end

    def recreate_huffman_codes(code_lengths)
        symbols_with_lengths = code_lengths.each_with_index.filter {|len, _| len > 0}.map {|len, s| [s, len] }.sort_by {|s, len| [len, s]}

        res = Array.new(code_lengths.size)
        prev_length = 0
        code = 0

        symbols_with_lengths.each do |s, len|
            code <<= (len - prev_length)
            res[s] = code.to_s(2).rjust(len, '0')
            code += 1
            prev_length = len
        end

        res
    end

    def decode_code_lengths_lengths(header)
        bits = header.bytes.map {|ch| ch.to_s(2).rjust(8, '0')}.join

        code_lengths = []

        (0..18).each do |e|
            start_bit = e * 3
            three_bits = bits[start_bit, 3]
            length = three_bits.to_i(2)
            code_lengths[e] = length
        end

        code_lengths
    end

    def decode_codes_lengths_with_position(header, codes_lengths_tree)
        bits = header.bytes.map {|b| b.to_s(2).rjust(8, '0')}.join

        tree_inv = codes_lengths_tree.each_with_index.to_h

        bit_pos = 0
        buf = ''

        res = []

        while bit_pos < bits.length && res.size < 256
            buf += bits[bit_pos]
            bit_pos += 1

            if tree_inv.key?(buf)
                symbol = tree_inv[buf]
                buf = ''

                case symbol
                when 16
                    extra_bits = bits[bit_pos, 2].to_i(2)
                    bit_pos += 2
                    repeats = 3 + extra_bits
                    repeats.times { res << res.last }
                when 17
                    extra_bits = bits[bit_pos, 3].to_i(2)
                    bit_pos += 3
                    repeats = 3 + extra_bits
                    repeats.times { res << 0 }
                when 18
                    extra_bits = bits[bit_pos, 7].to_i(2)
                    bit_pos += 7
                    repeats = 11 + extra_bits
                    repeats.times { res << 0 }
                else
                    res << symbol
                end
            end
        end

        [ res, bit_pos ]
    end

    def decode_body(body, codes_lengths, length)
        bits = body.bytes.map {|b| b.to_s(2).rjust(8, '0')}.join
        bytes = decode_with_tree(bits, codes_lengths, length)
        bytes.pack('C*').force_encoding('UTF-8')
    end

    def decode_with_tree(bits, tree, max_symbols = nil)
        tree_inv = tree.each_with_index.to_h

        res = []
        buf = ''

        bits.each_char do |bit|
            buf += bit

            if tree_inv.key? buf
                res << tree_inv[buf]
                buf = ''

                break if max_symbols && res.length >= max_symbols
            end
        end

        res
    end

    def decode_impl(message)
      length = message[0..3].unpack1('N')
      header1 = message[4..11]

      code_lengths_lengths = decode_code_lengths_lengths(header1)
      tree = recreate_huffman_codes(code_lengths_lengths)

      header2_start = message[12..]
      codes_lengths, pos = decode_codes_lengths_with_position(header2_start, tree)

      header2_length = (pos + 7) / 8
      body = message[(12 + header2_length)..]
      codes = recreate_huffman_codes(codes_lengths)

      decode_body(body, codes, length)
    end
  end
end

if ARGV[0] == 'encode'
  STDOUT.print SimpleDeflate.encode(STDIN.read)
elsif ARGV[0] == 'decode'
  STDOUT.print SimpleDeflate.decode(STDIN.read)
end

File: simple_deflate.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <algorithm>
#include <memory>
#include <optional>

std::string to_binary(uint64_t n, size_t width) {
    std::string result;
    for (int i = width - 1; i >= 0; --i) {
        result += ((n >> i) & 1) ? '1' : '0';
    }
    return result;
}

std::string pad_left(std::string s, size_t width, char fill) {
    if (s.length() >= width) return s;
    return std::string(width - s.length(), fill) + s;
}

std::string pad_right(std::string s, size_t width, char fill) {
    if (s.length() >= width) return s;
    return s + std::string(width - s.length(), fill);
}

std::vector<uint8_t> string_to_bytes(const std::string& s) {
    return std::vector<uint8_t>(s.begin(), s.end());
}

std::string bytes_to_string(const std::vector<uint8_t>& bytes) {
    return std::string(bytes.begin(), bytes.end());
}

struct TreeNode {
    std::optional<uint8_t> ch;
    int count;
    std::shared_ptr<TreeNode> left;
    std::shared_ptr<TreeNode> right;
    
    TreeNode(std::optional<uint8_t> c, int cnt) : ch(c), count(cnt), left(nullptr), right(nullptr) {}
};

std::shared_ptr<TreeNode> build_tree(const std::vector<uint8_t>& chars) {
    std::map<uint8_t, int> freq;
    for (auto c : chars) freq[c]++;
    
    std::vector<std::shared_ptr<TreeNode>> pq;
    for (auto [c, count] : freq) {
        pq.push_back(std::make_shared<TreeNode>(c, count));
    }
    
    auto cmp = [](const std::shared_ptr<TreeNode>& a, const std::shared_ptr<TreeNode>& b) {
        if (a->count != b->count) return a->count > b->count;
        return (a->ch.value_or(0)) < (b->ch.value_or(0));
    };
    std::sort(pq.begin(), pq.end(), cmp);
    
    while (pq.size() > 1) {
        auto right = pq.back(); pq.pop_back();
        auto left = pq.back(); pq.pop_back();
        
        if (left->count < right->count) std::swap(left, right);
        
        auto parent = std::make_shared<TreeNode>(std::nullopt, left->count + right->count);
        parent->left = left;
        parent->right = right;
        
        pq.push_back(parent);
        std::sort(pq.begin(), pq.end(), cmp);
    }
    
    return pq[0];
}

std::map<uint8_t, std::string> build_table(std::shared_ptr<TreeNode> tree) {
    struct QueueItem { std::shared_ptr<TreeNode> node; std::string code; };
    std::vector<QueueItem> q = {{tree, ""}};
    std::map<uint8_t, std::string> codes;
    
    while (!q.empty()) {
        auto e = q.front();
        q.erase(q.begin());
        
        if (!e.node->left && !e.node->right) {
            codes[e.node->ch.value()] = e.code;
            continue;
        }
        
        if (e.node->left) q.push_back({e.node->left, e.code + "0"});
        if (e.node->right) q.push_back({e.node->right, e.code + "1"});
    }
    
    return codes;
}

std::map<uint8_t, std::string> build_canonical_table(const std::map<uint8_t, std::string>& codes) {
    std::vector<std::pair<uint8_t, std::string>> sorted_codes(codes.begin(), codes.end());
    std::sort(sorted_codes.begin(), sorted_codes.end(), [](auto& a, auto& b) {
        if (a.second.length() != b.second.length()) return a.second.length() < b.second.length();
        return a.first < b.first;
    });
    
    std::map<uint8_t, std::string> canonical_codes;
    uint64_t curr_code = 0;
    size_t prev_length = 0;
    
    for (auto [ch, old_code] : sorted_codes) {
        curr_code <<= (old_code.length() - prev_length);
        canonical_codes[ch] = pad_left(to_binary(curr_code, old_code.length()), old_code.length(), '0');
        curr_code += 1;
        prev_length = old_code.length();
    }
    
    return canonical_codes;
}

std::vector<std::optional<std::string>> recreate_huffman_codes(const std::vector<int>& code_lengths) {
    std::vector<std::pair<int, int>> symbols_with_lengths;
    for (int i = 0; i < code_lengths.size(); ++i) {
        if (code_lengths[i] > 0) {
            symbols_with_lengths.push_back({i, code_lengths[i]});
        }
    }

    std::sort(symbols_with_lengths.begin(), symbols_with_lengths.end(), [](auto& a, auto& b) {
        if (a.second != b.second) return a.second < b.second;
        return a.first < b.first;
    });

    std::vector<std::optional<std::string>> res(code_lengths.size());
    int prev_length = 0;
    uint64_t code = 0;

    for (auto [s, len] : symbols_with_lengths) {
        code <<= (len - prev_length);
        res[s] = pad_left(to_binary(code, len), len, '0');
        code += 1;
        prev_length = len;
    }

    return res;
}

struct RLEItem {
    bool is_pair;
    int value;
    int extra;  // only used when is_pair == true

    RLEItem(int v) : is_pair(false), value(v), extra(0) {}
    RLEItem(int v, int e) : is_pair(true), value(v), extra(e) {}
};

std::vector<RLEItem> running_codes_lengths(const std::map<uint8_t, std::string>& codes) {
    // Create array of code lengths for bytes 0-255
    std::vector<int> codes_lengths(256, 0);
    for (auto [byte, code] : codes) {
        codes_lengths[byte] = code.length();
    }

    std::vector<RLEItem> result;
    size_t i = 0;

    while (i < codes_lengths.size()) {
        int len = codes_lengths[i];
        int run_length = 1;

        while (i + run_length < codes_lengths.size() && codes_lengths[i + run_length] == len) {
            run_length++;
        }

        i += run_length;

        if (len == 0) {
            while (run_length > 0) {
                if (run_length >= 11) {
                    int diff = std::min(run_length, 138);
                    result.push_back(RLEItem(18, diff - 11));
                    run_length -= diff;
                } else if (run_length >= 3) {
                    int diff = std::min(run_length, 10);
                    result.push_back(RLEItem(17, diff - 3));
                    run_length -= diff;
                } else {
                    for (int j = 0; j < run_length; ++j) result.push_back(RLEItem(0));
                    run_length = 0;
                }
            }
        } else if (len != 0 && run_length >= 3) {
            result.push_back(RLEItem(len));
            run_length -= 1;

            while (run_length >= 3) {
                int diff = std::min(run_length, 6);
                result.push_back(RLEItem(16, diff - 3));
                run_length -= diff;
            }
        }

        // Add any remaining values individually
        for (int j = 0; j < run_length; ++j) result.push_back(RLEItem(len));
    }

    return result;
}

std::string encode(const std::string& message) {
    auto bytes = string_to_bytes(message);
    auto tree = build_tree(bytes);
    auto codes = build_table(tree);
    auto canonical_codes = build_canonical_table(codes);

    auto codes_lengths = running_codes_lengths(canonical_codes);

    std::vector<uint8_t> raw_codes_lengths;
    for (auto& item : codes_lengths) {
        raw_codes_lengths.push_back(item.value);
    }

    auto codes_lengths_tree = build_tree(raw_codes_lengths);
    auto codes_lengths_codes = build_table(codes_lengths_tree);
    auto canonical_codes_lengths = build_canonical_table(codes_lengths_codes);

    std::string body_bits;
    for (auto c : bytes) {
        body_bits += canonical_codes[c];
    }
    body_bits = pad_right(body_bits, ((body_bits.length() + 7) / 8) * 8, '0');

    std::string encoded_body;
    for (size_t i = 0; i < body_bits.length(); i += 8) {
        uint8_t byte = 0;
        for (int j = 0; j < 8; ++j) {
            byte = (byte << 1) | (body_bits[i + j] - '0');
        }
        encoded_body += static_cast<char>(byte);
    }

    std::string encoded_codes_lengths;
    for (auto& c : codes_lengths) {
        if (c.is_pair) {
            int extra_bits = (c.value == 16) ? 2 : (c.value == 17) ? 3 : 7;
            encoded_codes_lengths += canonical_codes_lengths[c.value];
            encoded_codes_lengths += pad_left(to_binary(c.extra, extra_bits), extra_bits, '0');
        } else {
            encoded_codes_lengths += canonical_codes_lengths[c.value];
        }
    }

    std::string bit_string;
    for (int c = 0; c <= 18; ++c) {
        int len = canonical_codes_lengths.count(c) ? canonical_codes_lengths[c].length() : 0;
        bit_string += pad_left(to_binary(len, 3), 3, '0');
    }
    bit_string = pad_right(bit_string, ((bit_string.length() + 7) / 8) * 8, '0');

    std::string encoded_codes_lengths_tree;
    for (size_t i = 0; i < bit_string.length(); i += 8) {
        uint8_t byte = 0;
        for (int j = 0; j < 8; ++j) {
            byte = (byte << 1) | (bit_string[i + j] - '0');
        }
        encoded_codes_lengths_tree += static_cast<char>(byte);
    }

    encoded_codes_lengths = pad_right(encoded_codes_lengths, ((encoded_codes_lengths.length() + 7) / 8) * 8, '0');

    std::string encoded_codes_lengths_body;
    for (size_t i = 0; i < encoded_codes_lengths.length(); i += 8) {
        uint8_t byte = 0;
        for (int j = 0; j < 8; ++j) {
            byte = (byte << 1) | (encoded_codes_lengths[i + j] - '0');
        }
        encoded_codes_lengths_body += static_cast<char>(byte);
    }

    std::string result;

    uint32_t length = bytes.size();
    result += static_cast<char>((length >> 24) & 0xFF);
    result += static_cast<char>((length >> 16) & 0xFF);
    result += static_cast<char>((length >> 8) & 0xFF);
    result += static_cast<char>(length & 0xFF);

    result += encoded_codes_lengths_tree;
    result += encoded_codes_lengths_body;
    result += encoded_body;

    return result;
}

std::vector<int> decode_code_lengths_lengths(const std::string& header) {
    auto bytes = string_to_bytes(header);
    std::string bits;
    for (auto ch : bytes) {
        bits += pad_left(to_binary(ch, 8), 8, '0');
    }

    std::vector<int> code_lengths;
    for (int e = 0; e <= 18; ++e) {
        int start_bit = e * 3;
        std::string three_bits = bits.substr(start_bit, 3);
        int length = std::stoi(three_bits, nullptr, 2);
        code_lengths.push_back(length);
    }

    return code_lengths;
}

std::pair<std::vector<int>, int> decode_codes_lengths_with_position(const std::string& header, const std::vector<std::optional<std::string>>& codes_lengths_tree) {
    auto bytes = string_to_bytes(header);
    std::string bits;
    for (auto b : bytes) {
        bits += pad_left(to_binary(b, 8), 8, '0');
    }

    std::map<std::string, int> tree_inv;
    for (int i = 0; i < codes_lengths_tree.size(); ++i) {
        if (codes_lengths_tree[i].has_value()) {
            tree_inv[codes_lengths_tree[i].value()] = i;
        }
    }

    int bit_pos = 0;
    std::string buf;
    std::vector<int> res;

    while (bit_pos < bits.length() && res.size() < 256) {
        buf += bits[bit_pos];
        bit_pos++;

        if (tree_inv.count(buf)) {
            int symbol = tree_inv[buf];
            buf.clear();

            if (symbol == 16) {
                std::string extra_bits_str = bits.substr(bit_pos, 2);
                int extra_bits = std::stoi(extra_bits_str, nullptr, 2);
                bit_pos += 2;
                int repeats = 3 + extra_bits;
                for (int i = 0; i < repeats; ++i) res.push_back(res.back());
            } else if (symbol == 17) {
                std::string extra_bits_str = bits.substr(bit_pos, 3);
                int extra_bits = std::stoi(extra_bits_str, nullptr, 2);
                bit_pos += 3;
                int repeats = 3 + extra_bits;
                for (int i = 0; i < repeats; ++i) res.push_back(0);
            } else if (symbol == 18) {
                std::string extra_bits_str = bits.substr(bit_pos, 7);
                int extra_bits = std::stoi(extra_bits_str, nullptr, 2);
                bit_pos += 7;
                int repeats = 11 + extra_bits;
                for (int i = 0; i < repeats; ++i) res.push_back(0);
            } else {
                res.push_back(symbol);
            }
        }
    }

    return {res, bit_pos};
}

std::string decode(const std::string& message) {
    auto bytes = string_to_bytes(message);

    uint32_t length = (static_cast<uint32_t>(bytes[0]) << 24) |
                      (static_cast<uint32_t>(bytes[1]) << 16) |
                      (static_cast<uint32_t>(bytes[2]) << 8) |
                      static_cast<uint32_t>(bytes[3]);

    std::string header1 = message.substr(4, 8);

    auto code_lengths_lengths = decode_code_lengths_lengths(header1);
    auto tree = recreate_huffman_codes(code_lengths_lengths);

    std::string header2_start = message.substr(12);
    auto [codes_lengths, pos] = decode_codes_lengths_with_position(header2_start, tree);

    int header2_byte_length = (pos + 7) / 8;
    std::string body = message.substr(12 + header2_byte_length);

    auto codes = recreate_huffman_codes(codes_lengths);

    auto body_bytes = string_to_bytes(body);
    std::string bits;
    for (auto b : body_bytes) {
        bits += pad_left(to_binary(b, 8), 8, '0');
    }

    std::map<std::string, int> tree_inv;
    for (int i = 0; i < codes.size(); ++i) {
        if (codes[i].has_value()) {
            tree_inv[codes[i].value()] = i;
        }
    }

    std::vector<uint8_t> res;
    std::string buf;

    for (char bit : bits) {
        buf += bit;

        if (tree_inv.count(buf)) {
            res.push_back(tree_inv[buf]);
            buf.clear();

            if (res.size() >= length) {
                break;
            }
        }
    }

    return bytes_to_string(res);
}

int main(int argc, char* argv[]) {
    if (argc != 2 || (std::string(argv[1]) != "encode" && std::string(argv[1]) != "decode")) {
        std::cerr << "Usage: " << argv[0] << " encode|decode < input > output\n";
        return 1;
    }

    std::string input((std::istreambuf_iterator<char>(std::cin)), std::istreambuf_iterator<char>());

    if (std::string(argv[1]) == "encode") {
        std::cout << encode(input);
    } else {
        std::cout << decode(input);
    }

    return 0;
}


